### PR TITLE
provider/aws: Add support for Network Loadbalancers

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #1618

In terraform, we had the idea of an alb. In AWS this doesn't exist. ALBs
are actually Load balancers of type `application`

Therefore, the first part of this PR adds a new parameter to ALBs called
`load_balancer_type`. We default this to `application` to follow the
same idea as the current behaviour

The next part of the PR will then change the idea of an alb -> lb

In order to preserve backwards compatibility, we have added another
resource name to the same schema type. This means we effectively have an
alias of aws_alb and aws_lb. This includes updating *all* of the tests
to make sure and remove the idea of ALB and rename to LB and then we
will add a check to make sure we can still check that an ALB can be
created in the old resource

### CODE REVIEW

**Comment by stack72** [comment]
Not sure what happened in #1629 - my remote branch got deleted so i pushed a new version and it closed the PR

**Comment by radeksimko** [comment]
> When you are happy it's the same, I will make all of the changes you asked for

I'm pretty confident it's the same, based on patch I was able to pull from my local branch when reviewing the old PR: https://gist.github.com/radeksimko/f0e953ba5bae73c05732648d5bafc032

So feel free to make the changes.

**Comment by stack72** [comment]
Hi @radeksimko 

Ok, I made the docs changes. The issue with this line:

```
if d.Get("load_balancer_type").(string) != "network" && d.HasChange("idle_timeout") {
```

We set a default value for `idle_timeout` and the way that Create calls Update func, we don't want to set that value if the type of loadbalancer is `network`

Does this make more sense? Do you think there is a better way to handle this?

P.

File: aws/resource_aws_lb_listener.go [hunk 1]
```diff
@@ -252,15 +252,15 @@ func resourceAwsAlbListenerDelete(d *schema.ResourceData, meta interface{}) erro
...
 	if port < 1 || port > 65536 {
 		errors = append(errors, fmt.Errorf("%q must be a valid port number (1-65536)", k))
 	}
 	return
 }
 
-func validateAwsAlbListenerProtocol(v interface{}, k string) (ws []string, errors []error) {
+func validateAwsLbListenerProtocol(v interface{}, k string) (ws []string, errors []error) {
 	value := strings.ToLower(v.(string))
 	if value == "http" || value == "https" {
```
Comment (jwinter) [inline aws/resource_aws_lb_listener.go]: Network Load Balancers only support TCP listeners, so I think line 265 needs to be updated to allow for a value of "tcp".

https://github.com/aws/aws-sdk-go/blob/master/service/elbv2/api.go#L3294-L3296

Comment (stack72) [reply aws/resource_aws_lb_listener.go]: @jwinter you are indeed correct - just pushed this change!

**Comment by stack72** [comment]
@radeksimko I agree - let's certainly chat about this and see if we can solve this in another fashion as a followup

**Comment by radeksimko** [comment]
@handlerbot I think we first need to implement mechanism for deprecating resources. Then we're _actually_ going to mark those old names as deprecated and they'll be reported as such when used in configs. This is work to be done in core schema package. Until that's done, aliases will remain in place as we want to avoid surprises.

There is no migration tool/mechanism available at this point for any deprecations, although we did discuss the idea of having one internally within the team. If we don't have one at the time of real deprecation then yes, it will require `terraform state mv` surgery + config changes. It is highly likely that we'll present a semi-automated way leveraging unix tools like `find`/`sed` etc. but we're talking future here. I'm aware of how inconvenient would such migration be.

I don't have any timelines to share around any of the above, but I'm pretty sure that from introducing resource deprecation (in the schema package) it would take at least a couple of months, or rather half a year until the aliases are removed.

File: aws/resource_aws_lb_target_group.go [hunk 2]
```diff
@@ -410,15 +410,15 @@ func validateAwsAlbTargetGroupHealthCheckProtocol(v interface{}, k string) (ws [
...
 	if port < 1 || port > 65536 {
 		errors = append(errors, fmt.Errorf("%q must be a valid port number (1-65536)", k))
 	}
 	return
 }
 
-func validateAwsAlbTargetGroupProtocol(v interface{}, k string) (ws []string, errors []error) {
+func validateAwsLbTargetGroupProtocol(v interface{}, k string) (ws []string, errors []error) {
 	protocol := strings.ToLower(v.(string))
 	if protocol == "http" || protocol == "https" {
```
Comment (cbarbara-okta) [inline aws/resource_aws_lb_target_group.go]: this needs to allow "tcp", just like your change to `validateAwsLbListenerProtocol`. "tcp" is a valid network load balancer target group protocol.

**Pull request merged**